server/src/main/java/org/opensearch/cluster/metadata/QueryGroup.java

```diff
@@ -0,0 +1,324 @@
+            Objects.requireNonNull(resource.getKey(), "resourceName can't be null");
+            Objects.requireNonNull(threshold, "resource limit threshold for" + resource.getKey().getName() + " : can't be null");
+
+            if (Double.compare(threshold, 1.0) > 0) {
+                throw new IllegalArgumentException("resource value should be less than 1.0");
+            }
+        }
+    }
+
+    /**
+     * @param builder
+     * @param params
+     * @return
+     * @throws IOException
+     */
+    @Override
+    public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startObject();
+        builder.field("_id", _id);
+        builder.field("name", name);
+        builder.field("resiliency_mode", resiliencyMode.getName());
+        builder.field("updatedAt", updatedAtInMillis);
+        // write resource limits
+        builder.startObject("resourceLimits");
+        for (Map.Entry<ResourceType, Object> resourceLimit : resourceLimits.entrySet()) {
+            builder.field(resourceLimit.getKey().getName(), resourceLimit.getValue());
+        }
+        builder.endObject();
+
+        builder.endObject();
+        return builder;
+    }
+
+    public static QueryGroup fromXContent(final XContentParser parser) throws IOException {
+        if (parser.currentToken() == null) { // fresh parser? move to the first token
+            parser.nextToken();
+        }
+
+        Builder builder = builder();
+
+        XContentParser.Token token = parser.currentToken();
+
+        if (token != XContentParser.Token.START_OBJECT) {
+            throw new IllegalArgumentException("Expected START_OBJECT token but found [" + parser.currentName() + "]");
+        }
+
+        String fieldName = "";
+        // Map to hold resources
+        final Map<ResourceType, Object> resourceLimits = new HashMap<>();
+        while ((token = parser.nextToken()) != null) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                fieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if (fieldName.equals("_id")) {
+                    builder._id(parser.text());
+                } else if (fieldName.equals("name")) {
+                    builder.name(parser.text());
+                } else if (fieldName.equals("resiliency_mode")) {
+                    builder.mode(parser.text());
+                } else if (fieldName.equals("updatedAt")) {
+                    builder.updatedAt(parser.longValue());
+                } else {
+                    throw new IllegalArgumentException(fieldName + " is not a valid field in QueryGroup");
+                }
+            } else if (token == XContentParser.Token.START_OBJECT ) {
+
+                if (!fieldName.equals("resourceLimits")) {
+                    throw new IllegalArgumentException("was expecting the { " + " but found " + token);
```

Comment:
nit: The exception message looks incorrect, doesn't mention resourceLimits field at all